src/Symfony/Component/HttpClient/NoPrivateNetworkHttpClient.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpClient\Exception\InvalidArgumentException;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpFoundation\IpUtils;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * Decorator that block requests to private network by default.
+ *
+ * @author Hallison Boaventura <hallisonboaventura@gmail.com>
+ */
+final class NoPrivateNetworkHttpClient implements HttpClientInterface, LoggerAwareInterface
+{
+    use HttpClientTrait;
+
+    const IPV4_PRIVATE_SUBNETS = [
+        '127.0.0.0/8',
+        '10.0.0.0/8',
+        '192.168.0.0/16',
+        '172.16.0.0/12',
+        '169.254.0.0/16',
+        '0.0.0.0/8',
+        '240.0.0.0/4',
+    ];
+
+    const IPV6_PRIVATE_SUBNETS = [
+        '::1/128',
+        'fc00::/7',
+        'fe80::/10',
+        '::ffff:0:0/96',
+        '::/128',
+    ];
+
+    /**
+     * @var HttpClientInterface
+     */
+    private $client;
+
+    /**
+     * Subnets in CIDR notation that IpUtils will make use of.
+     *
+     * @var string|array|null
+     */
+    private $subnets;
+
+    /**
+     * Constructor.
+     *
+     * @param HttpClientInterface $client  A HttpClientInterface instance.
+     * @param string|array|null   $subnets String or array of subnets using CIDR notation that will be used by IpUtils.
+     *                                     If its value is null, then default private subnets will be used.
+     */
+    public function __construct(HttpClientInterface $client, $subnets = null)
+    {
+        $this->client = $client;
+
+        if (!(\is_string($subnets) || \is_array($subnets) || null === $subnets)) {
+            throw new InvalidArgumentException('$subnets argument must be string, array or null.');
+        }
+
+        $this->subnets = $subnets;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        $subnets = $this->subnets;
+
+        $lastPrimaryIp = '';
+        $onProgress = $options['on_progress'] ?? null;
```

Comment:
doing this should allow more accurate error reporting when the progress callback is broken:
```
if (!\is_callable($options['on_progress'] ?? 'var_dump')) {
    return $this->client->request($method, $url, $options);
}
```

Reply:
Sorry, Nicolas, I didn't get this part. If we return when user pass non-callable on on_progress, then we never parse his IP. You mean use var_dump by default if on_progress is non-callable or is missing? Then we can securelly call $onProgress because it will always is a callable?

Reply:
If the option is non-null but is not a callable, and error will be raised so there is nothing to do.